src/ControlSystem/FunctionVsTime.hpp

```diff
@@ -0,0 +1,169 @@
+    result[i][MaxDeriv] = derivs[i][MaxDeriv];
+  }
+
+  return result;
+}
+
+template <size_t MaxDeriv>
+void FunctionVsTimeNthDeriv<MaxDeriv>::update(
+    double time_of_update,
+    const std::vector<double>& updated_max_deriv) noexcept {
+
+  if (time_of_update <= deriv_info_.back().time) {
+    ERROR("FunctionVsTimeNthDeriv::update: "
+          "t must be increasing from call to call.");
+  }
+
+  // get the current values, before updating the `MaxDeriv'th deriv
+  value_type func = this->operator()(time_of_update);
+
+  if (updated_max_deriv.size() != func.size()) {
+    ERROR("FunctionVsTimeNthDeriv::update: "
+          "the number of components trying to be updated ("
+          << updated_max_deriv.size() << ") does "
+          "not match the number of components ("
+          << func.size() << ") in the fvt.");
+  }
+
+  // update the MaxDerivs
+  for (size_t i = 0; i < updated_max_deriv.size(); i++) {
+    func[i][MaxDeriv] = updated_max_deriv[i];
+  }
+  deriv_info_.emplace_back(std::move(time_of_update),std::move(func));
+}
+
+// Return iterator pointing to k, such that times[k] <= t < times[k+1].
+// For t < times[0], return times.begin()
+// For t > times[times.size()-1] : return prev(times.end(),1)
+template <size_t MaxDeriv>
+auto FunctionVsTimeNthDeriv<MaxDeriv>::closest_time_iter(
+    const double t) const noexcept {
+  // this function now assumes that times is sorted, which is enforced
+  // by the update function.
+
+  // returns iterator pointing to the first element that is > than value.
+  // returns times.end() if no element > than value.
+  // if value < min(times), then returns times.begin() index.
+  auto it = std::upper_bound(deriv_info_.begin(), deriv_info_.end(), t,
+                       [](double t0, const DerivInfo& d) {return d.time > t0;});
+
+  // this section needs to change if extrapolation is not allowed.
+  // This dependes on whether the mode is "Replay" or "Evolution",
+  // which will be implemented in the future.
+  bool extrapolate = true; // will eventually be set according to mode.
+  const double eps = 1.0e-15;
+
+  if (it == deriv_info_.end()){
+    // t is greater than all elements of times
+    const auto last_element = std::prev(it,1);
+    const auto last_time = last_element->time;
+    if(extrapolate or
+       t-last_time<(std::max(std::abs(t),std::abs(last_time))+1.0)*eps){
+      // extrapolation allowed or within round-off of last t in times
+      return last_element;
+    } else{
+      // no extrapolation allowed and beyond roundoff
+      ERROR("FunctionVsTimeNthDeriv::closest_time_iter: "
+            << "requested time " << t
+            << " beyond maximum time " << last_time
+            <<  " of times and extrapolation is prohibited.");
+      // the above error cannot be tested until extrapolate is implemented
+    }
+  } else if(it == deriv_info_.begin()){
+    // all elements are greater than t
+    if(it->time-t<(std::max(std::abs(t),std::abs(it->time))+1.0)*eps){
```

Comment:
reverse the logic of the bool, move the ERROR into the if, don't use an else, and then return it;